File: Python3/0316_remove_duplicate_letters.py
# https://leetcode.cn/problems/remove-duplicate-letters/

class Solution:
    def removeDuplicateLetters(self, s):
        """
        Inputs:
            s: string
        Outputs:
            res: string
        """

        counts = [0 for _ in range(26)]
        check = [0 for _ in range(26)]
        for ch in s:
            counts[ord(ch) - ord('a')] = counts[ord(ch) - ord('a')] + 1
        
        res = ""
        for ch in s:
            idx = ord(ch) - ord('a')
            if check[idx] == 0:
                while (res != "") and (res[-1] > ch):
                    if counts[ord(res[-1]) - ord('a')] > 0:
                        check[ord(res[-1]) - ord('a')] = 0
                        res = res[: -1]
                    else:
                        break

                res = res + ch
                check[idx] = 1
            counts[idx] = counts[idx] - 1
        return res

File: Cpp/0316_remove_duplicate_letters.cpp
// https://leetcode.cn/problems/remove-duplicate-letters/

class Solution {
public:
    string removeDuplicateLetters(string s) {
        vector<int> counts(26), check(26);
        for (char ch: s) {
            counts[ch-'a'] = counts[ch-'a'] + 1;
        }

        string res;
        for (char ch: s) {
            int idx = ch - 'a';
            if (!check[idx]) {
                while ((!res.empty()) && (res.back() > ch)) {
                    if (counts[res.back()-'a'] > 0) {
                        check[res.back()-'a'] = 0;
                        res.pop_back();
                    }
                    else {
                        break;
                    }
                }
                res.push_back(ch);
                check[idx] = 1;
            }
            counts[idx] = counts[idx] - 1;
        }
        return res;
    }
};
